pkg/controller/azure/provider/resourcegroup.go

```diff
@@ -0,0 +1,148 @@
+	r := &ResourceGroupReconciler{
+		Client:     mgr.GetClient(),
+		scheme:     mgr.GetScheme(),
+		kubeclient: kubernetes.NewForConfigOrDie(mgr.GetConfig()),
+		recorder:   mgr.GetRecorder(controllerNameRG),
+	}
+	r.create = r._create
+	return r
+}
+
+// AddRG creates a new Resource Group Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func AddRG(mgr manager.Manager) error {
+	return addRG(mgr, newResourceGroupReconciler(mgr))
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func addRG(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerNameRG, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Provider
+	err = c.Watch(&source.Kind{Type: &v1alpha1.ResourceGroup{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// fail - helper function to set fail condition with reason and message
+func (r *ResourceGroupReconciler) fail(instance *v1alpha1.ResourceGroup, reason, msg string) (reconcile.Result, error) {
+	instance.Status.UnsetAllConditions()
+	instance.Status.SetFailed(reason, msg)
+	return resultRequeue, r.Update(context.TODO(), instance)
+}
+
+func (r *ResourceGroupReconciler) _create(client *azureclient.Client, name string, location string) error {
+	return azureclient.CreateOrUpdateGroup(client, name, location)
+}
+
+// Reconcile reads that state of the cluster for a ResourceGroup object and makes changes based on the state read
+// and what is in the ResourceGroup.Spec
+func (r *ResourceGroupReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("reconciling %s: %v", v1alpha1.ProviderKindAPIVersion, request)
+	instance := &v1alpha1.ResourceGroup{}
+
+	// Fetch the ResourceGroup instance
+	err := r.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		log.Printf("failed to get object at start of reconcile loop: %+v", err)
+		return reconcile.Result{}, err
+	}
+
+	// look up the provider information for this instance
+	provider := &v1alpha1.Provider{}
+	providerNamespacedName := apitypes.NamespacedName{
+		Namespace: instance.Namespace,
+		Name:      instance.Spec.ProviderRef.Name,
+	}
+	if err := r.Get(ctx, providerNamespacedName, provider); err != nil {
+		return r.fail(instance, errorFetchingAzureProvider, fmt.Sprintf("failed to get provider on behalf of resource group %+v: %+v", providerNamespacedName, err))
+	}
+
+	// Retrieve azure client from the given provider config
+	azureClient, err := azureclient.NewClient(provider, r.kubeclient)
```

Comment:
i have a very dim memory of the controller-runtime client not doing something I needed with secrets, which resulted in our controllers also carrying around a `kubernetes.Interface` reference.  That definitely seems redundant and it's worth looking into that again to see if we can get rid of that duplication.  It would be great to use just the `client.Client` type in our controllers.  I've opened #319 to track this.